### PR TITLE
LikeDislike videos

### PR DESCRIPTION
**Pull request by richardARPANET**
## Description

- Backend API
- Frontend component, progressive rendering

Missing:

- JS tests. are they even needed when we manually test everything anyway?

## Issue Link

(part of) 
https://github.com/VeemsHQ/veems/issues/56

## Screenshots (if appropriate):

https://user-images.githubusercontent.com/38049750/106060486-0ce7d380-60ec-11eb-9a53-c335b38a944e.mp4

## Checklist:

- [x] My code includes automated tests (if this change warrants them).

### CODE REVIEW

File: react-components/src/components/VideoLikeDislikeWidget/index.js [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+  const [isLiked, setIsLiked] = useState(props.isLiked);
+  const [likesDislikesPercentage, setLikesDislikesPercentage] = useState(props.likesDislikesPercentage);
+
+  const updateStateFromApiResponse = async (requestPromise) => {
+    const { response, data } = await requestPromise;
+    if (response?.status === 403) {
+      setApiError('You need to login to do that.')
+    } else if (response?.status > 400) {
+      setApiError('Something went wrong, please try again.')
+    } else {
```
Comment (richardARPANET) [inline react-components/src/components/VideoLikeDislikeWidget/index.js:31]: looks like something common that should be handled in the api call layer but not sure how to reorg all the code+state to make that work.

File: react-components/src/components/VideoLikeDislikeWidget/VideoLikeDislikeWidget.js [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+      </div>
+    )
+  }
+
+  const getLikeControlTextColor = () => {
+    if (isLiked === true) {
+      return 'text-primary';
+    } else {
+      return 'text-muted';
+    }
```
Comment (richtier) [inline react-components/src/components/VideoLikeDislikeWidget/VideoLikeDislikeWidget.js:121]: if you like terseness you might like `return isLiked ? 'text-primary' : 'text-muted'`

File: react-components/src/components/VideoLikeDislikeWidget/index.js [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+      setLikesCount(data?.likes_count)
+      setDislikesCount(data?.dislikes_count)
+      setIsLiked(data?.is_like);
+      setLikesDislikesPercentage(data?.likesdislikes_percentage);
```
Comment (richtier) [inline react-components/src/components/VideoLikeDislikeWidget/index.js:35]: each set call will result in re-render. Not found a good way to solve for that other than using a dict for the values like
```
setVideoData({
    likeCount: ...
    dislikeCount: ...
    isLiked: ...
    percentage: ...
})
```
something to consider if/when doing performance optimization

File: react-components/src/components/VideoLikeDislikeWidget/VideoLikeDislikeWidget.js (hunk 2 above)
Comment (richardARPANET) [reply react-components/src/components/VideoLikeDislikeWidget/VideoLikeDislikeWidget.js:121]: :ok_hand:

File: templates/media/video.html [hunk 4]
```diff
@@ -8,29 +9,46 @@
...
+                                class="btn btn-sm d-flex align-items-center {% if video.authenticated_user_data.has_liked_video == True %}text-primary{% else %}text-muted{% endif %}"><i
+                                    class="small material-icons align-middle">thumb_up_alt</i><span
+                                    class="ml-2">{{video.likes_count}}</span></a>
+                            <a href="#" title="I dislike this"
+                                class="btn btn-sm d-flex align-items-center  {% if video.authenticated_user_data.has_liked_video == False %}text-primary{% else %}text-muted{% endif %}"><i
+                                    class="small material-icons align-middle">thumb_down_alt</i><span
+                                    class="ml-2">{{video.dislikes_count}}</span></a>
+                        </div>
+                        <div class="progress likedislike-progress">
+                            <div class="progress-bar {% if video.authenticated_user_data.has_liked_video %}bg-primary{% else %}bg-muted{% endif %}"
```
Comment (richtier) [inline templates/media/video.html:30]: if you're in terse gang:
```
{{  video.authenticated_user_data.has_liked_video|yesno:'bg-primary,bg-muted' }}
```

File: react-components/src/components/VideoLikeDislikeWidget/index.js (hunk 3 above)
Comment (richardARPANET) [reply react-components/src/components/VideoLikeDislikeWidget/index.js:35]: will fix that.

**Pull request merged**